WooCommerce/Classes/CardPresentPayments/CardPaymentsAdaptor.swift

```diff
@@ -0,0 +1,334 @@
+                        continuation.resume(returning: CardPresentPaymentResult.success(order))
+                    }
+                }
+            } onCancel: {
+                // TODO: cancel any in-progress discovery, connection, or payment.
+            }
+        }
+        self.paymentTask = paymentTask
+
+        return await paymentTask.value
+    }
+
+    func cancelPayment() {
+        paymentTask?.cancel()
+        paymentScreenEventSubject.send(nil) // This removes any otherwise-presented UI
+    }
+
+    @MainActor
+    func createTestOrder() async throws -> Order {
+        return try await withCheckedThrowingContinuation { continuation in
+            let action = OrderAction.createSimplePaymentsOrder(siteID: siteID,
+                                                               status: .pending,
+                                                               amount: "1.00",
+                                                               taxable: false) { result in
+                continuation.resume(with: result)
+            }
+            ServiceLocator.stores.dispatch(action)
+        }
+    }
+
+    enum CardPaymentsAdaptorError: Error, CardPaymentErrorProtocol {
+        var retryApproach: CardPaymentRetryApproach {
+            .restart
+        }
+
+        case unknownPaymentError(underlyingError: Error)
+    }
+}
+
+extension CardPresentPaymentsAdaptor: CardPresentPaymentAlertsPresenting {
+
+    func present(viewModel: CardPresentPaymentsModalViewModel) {
+        paymentScreenEventSubject.send(.presentAlert(viewModel))
+    }
+
+    func foundSeveralReaders(readerIDs: [String], connect: @escaping (String) -> Void, cancelSearch: @escaping () -> Void) {
+        paymentScreenEventSubject.send(.presentReaderList(readerIDs))
+        // the button actions here might need to be communicated... or we could expose them on the adaptor somehow.
+    }
+
+    func updateSeveralReadersList(readerIDs: [String]) {
+        paymentScreenEventSubject.send(.presentReaderList(readerIDs))
+    }
+
+    func dismiss() {
+        paymentScreenEventSubject.send(nil)
+        // TODO: Decide whether we really need this
+    }
+}
+
+import Combine
+
+/// This is really a re-implementation of the CardPresentPaymentsOnboardingPresenter, as it needs to take the calls to `showOnboardingIfRequired` and
+/// route the output to a SwiftUI view for display, rather than directly displaying on the viewController that's passed in.
+final class CardPresentPaymentsOnboardingPresenterAdaptor: CardPresentPaymentsOnboardingPresenting {
+    private let stores: StoresManager
+
+    private let onboardingUseCase: CardPresentPaymentsOnboardingUseCase
+
+    private let readinessUseCase: CardPresentPaymentsReadinessUseCase
+
+    private let onboardingViewModel: InPersonPaymentsViewModel
+
+    private var readinessSubscription: AnyCancellable?
+
+    var onboardingScreenViewModelPublisher: AnyPublisher<InPersonPaymentsViewModel?, Never>
```

Comment:
non-blocking suggestion: this is pre-existing and unrelated to this PR/experiment, just wanted to mention that `InPersonPaymentsViewModel` sounds a bit too generic to be used for an "onboarding" event type. If it's mostly for onboarding, the name could probably indicate that. From a quick look, this view model seems to do a bunch of things internally. It would make understanding how different classes play together easier.

Reply:
Yeah it's annoying for me too.

Renaming it before we start seems like a reasonable change to the existing code – it's just a choice that was obviously made very early in the IPP journey, and we've outgrown it.

Reply:
> nit: this could probably be `let` or `private(set)` at least

I don't think it can, if you want to bind to it... the setter needs to be public for that IIRC. Again, I could be confused and will check when I do it for real.